sdk/iot/hub/samples/src/paho_iot_hub_sas_telemetry_example.c

```diff
@@ -0,0 +1,424 @@
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "IoT Hub Device SAS Key Expiration (Hours)",
+      ENV_IOT_HUB_SAS_KEY_DURATION,
+      "2",
+      false,
+      iot_hub_sas_expiration,
+      &iot_hub_sas_expiration));
+
+  az_span iot_hub_sas_key_expiration_span
+      = az_span_init(az_span_ptr(iot_hub_sas_expiration), az_span_size(iot_hub_sas_expiration));
+  AZ_RETURN_IF_FAILED(az_span_atou32(iot_hub_sas_key_expiration_span, &iot_hub_sas_key_expiration));
+
+  iot_hub_sas_key_span = AZ_SPAN_FROM_BUFFER(iot_hub_sas_key);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "IoT Hub Device SAS Key",
+      ENV_IOT_HUB_SAS_KEY,
+      NULL,
+      true,
+      iot_hub_sas_key_span,
+      &iot_hub_sas_key_span));
+
+  az_span device_id_span = AZ_SPAN_FROM_BUFFER(device_id);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Device ID", ENV_DEVICE_ID, "", false, device_id_span, &device_id_span));
+
+  az_span iot_hub_hostname_span = AZ_SPAN_FROM_BUFFER(iot_hub_hostname);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "IoT Hub Hostname",
+      ENV_IOT_HUB_HOSTNAME,
+      "",
+      false,
+      iot_hub_hostname_span,
+      &iot_hub_hostname_span));
+
+  // Paho requires that the MQTT endpoint be of the form ssl://<HUB ENDPOINT>:8883
+  AZ_RETURN_IF_FAILED(
+      create_mqtt_endpoint(mqtt_endpoint, (int32_t)sizeof(mqtt_endpoint), iot_hub_hostname_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      &client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      NULL));
+
+  return AZ_OK;
+}
+
+static int generate_sas_key()
+{
+  az_result res;
+
+  // Create the POSIX expiration time from input hours
+  uint32_t sas_expiration = get_epoch_expiration_time(iot_hub_sas_key_expiration);
+
+  // Decode the base64 encoded SAS key to use for HMAC signing
+  az_span decoded_key_span;
+  if (az_failed(
+          res = sample_base64_decode(
+              iot_hub_sas_key_span, AZ_SPAN_FROM_BUFFER(sas_b64_decoded_key), &decoded_key_span)))
+  {
+    printf("Could not decode the SAS key: return code %d\n", res);
+    return res;
+  }
+
+  // Get the signature which will be signed with the decoded key
+  az_span sas_signature_span;
+  if (az_failed(
+          res = az_iot_hub_client_sas_get_signature(
+              &client,
+              sas_expiration,
+              AZ_SPAN_FROM_BUFFER(sas_signature_buf),
+              &sas_signature_span)))
+  {
+    printf("Could not get the signature for SAS key: return code %d\n", res);
+    return res;
+  }
+
+  // HMAC-SHA256 encrypt the signature with the decoded key
+  az_span hmac256_encrypted_span = AZ_SPAN_FROM_BUFFER(sas_signature_hmac_encoded_buf);
+  if (az_failed(
+          res = sample_hmac_sha256_encrypt(
```

Comment:
>sample_hmac_sha256_encrypt [](start = 16, length = 26)

we are only doing hmac with sha2 hash .... can we remove encrypt? It make you think that you are encrypting sha2 hmac

Reply:
how about `sample_hmac_sha256_sign`. Trying to be as descriptive as possible to make it easy for people to understand and port elsewhere.

Reply:
sample_hmac_sha256 is good enough

---
In reply to: [436079140](https://github.com/Azure/azure-sdk-for-c/pull/808#discussion_r436079140) [](ancestors = 436079140)